cli/src/commands/extensions/mod.rs

```diff
@@ -0,0 +1,113 @@
+                .arg(arg!([NAME]).required(true)),
+        )
+        .subcommand(Command::new("list").about("List installed extensions"))
+}
+
+/// Generate the subcommands for each extension.
+/// TODO add tests.
+pub fn extensions_subcommands(command: Command<'_>) -> Command<'_> {
+    let extensions = match list_extensions() {
+        Ok(extensions) => extensions,
+        Err(e) => {
+            error!("Couldn't list extensions: {}", e);
+            return command;
+        }
+    };
+
+    let names = command
+        .get_subcommands()
+        .map(|n| n.get_name().to_string())
+        .collect::<HashSet<_>>();
+
+    extensions
+        .into_iter()
+        .filter(|ext| !names.contains(ext.name()))
+        .fold(command, |command, ext| {
+            command.subcommand(Command::new(ext.name()))
+        })
+}
+
+/// Entry point for the `extensions` subcommand.
+pub async fn handle_extensions(matches: &ArgMatches) -> CommandResult {
+    if let Some(matches) = matches.subcommand_matches("add") {
+        subcmd_add_extension(matches.value_of_t("PATH")?).await
+    } else if let Some(matches) = matches.subcommand_matches("remove") {
+        subcmd_remove_extension(matches.value_of("NAME").unwrap()).await
+    } else {
+        // also covers the `list` case
+        subcmd_list_extensions().await
+    }
+}
+
+/// Handle the `extension add` subcommand path.
+/// Add the extension from the specified path.
+pub async fn subcmd_add_extension(path: PathBuf) -> CommandResult {
+    let extension = Extension::try_from(path)?;
+    info!("Installing extension {}...", extension.name());
+
+    extension.install()?;
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+/// Handle the `extension remove` subcommand path.
+/// Remove the extension named as specified.
+pub async fn subcmd_remove_extension(name: &str) -> CommandResult {
+    // let extension = Extension::try_from(extensions_path()?.join(name))?;
+    let extension = Extension::load(name)?;
+
+    extension.uninstall()?;
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+/// Handle the `extension` / `extension list` subcommand path.
+/// List installed extensions.
+pub async fn subcmd_list_extensions() -> CommandResult {
+    let extensions = list_extensions()?;
+
+    if extensions.is_empty() {
+        println!("No extensions are currently installed.");
+    } else {
+        extensions.into_iter().for_each(|ext| {
+            println!("{:20}   {}", ext.name(), ext.description().unwrap_or(""));
+        });
+    }
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+// Return a list of installed extensions. Filter out invalid extensions instead of exiting early.
+fn list_extensions() -> Result<Vec<Extension>> {
+    let extension_path = extensions_path()?;
+
+    if !extension_path.exists() {
+        return Ok(Vec::new());
+    }
+
+    Ok(std::fs::read_dir(extension_path)?
+        .filter_map(|d| Extension::try_from(d.map(|d| d.path()).ok()?).ok())
```

Comment:
Is the `d` here supposed to stand for `direntry`? I'd honestly prefer `e` because I initially thought this was for `directory` which is inaccurate.

I'd also be fine with more verbosity here, even if it causes the closure to be multi-line.